selfdrive/pandad/pandad.py

```diff
@@ -62,12 +63,21 @@ def flash_panda(panda_serial: str) -> Panda:
 
 
 def main() -> NoReturn:
+  # override manager's immediate SystemExit to allow pandad to exit gracefully
+  def signal_handler(signum, frame):
+    nonlocal do_exit
+    do_exit = True
+    cloudlog.warning(f"caught signal {signum}, exiting")
+
+  do_exit = False
+  signal.signal(signal.SIGTERM, signal_handler)
+
```

Comment:
When creating a [new Python process with `multiprocessing.Process`](https://github.com/commaai/openpilot/blob/e8857a6713db6b8b3b8cd67aca1b8c046a4dccd9/system/manager/process.py#L224), it inherits the parent's signal handler. For manager, that is to cause a `SystemExit` immediately, which SIGKILLs pandad.cc before it has a chance to react to the SIGTERM and close the relay.

https://github.com/commaai/openpilot/blob/e8857a6713db6b8b3b8cd67aca1b8c046a4dccd9/system/manager/manager.py#L183-L184